src/com/rcv/Main.java

```diff
@@ -164,31 +163,84 @@ static String executeTabulation(ElectionConfig config) {
+        boolean encounteredProblemForThisSource = false;
+
+        if (source.filePath == null || source.filePath.isEmpty()) {
+          Logger.severe(
+              "Invalid source file: missing filePath"
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (source.firstVoteColumnIndex == null || source.firstVoteColumnIndex < 0) {
+          Logger.severe(
+              "Invalid source file: missing or invalid firstVoteColumnIndex: %s",
+              source.filePath
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (
+            config.isTabulateByPrecinctEnabled() &&
+            (source.precinctColumnIndex == null || source.precinctColumnIndex < 0)
+        ) {
+          Logger.severe(
+              "Invalid source file: missing or invalid precinctColumnIndex when " +
+                  "tabulateByPrecinct is enabled: %s",
+              source.filePath
+          );
+          encounteredProblemForThisSource = true;
+        }
+
+        if (!encounteredProblemForThisSource) {
+          // reader: read input file into a list of cast vote records
+          CVRReader reader = new CVRReader(
+              config,
+              source.filePath,
+              source.firstVoteColumnIndex,
+              source.idColumnIndex,
+              source.precinctColumnIndex
+          );
+          // the CVRs parsed from this source
+          List<CastVoteRecord> cvrs = reader.parseCVRFile();
+          if (cvrs.isEmpty()) {
+            Logger.severe("Source file contains no CVRS: %s", source.filePath);
+            encounteredProblemForThisSource = true;
+          }
+          // add records to the master list
+          castVoteRecords.addAll(cvrs);
+        }
+
+        if (encounteredProblemForThisSource) {
+          numInvalidSources++;
+        }
+      }
+      if (numInvalidSources > 0) {
+        Logger.severe("Encountered %d invalid source(s)", numInvalidSources);
+      } else {
+        Logger.info("Read %d cast vote records", castVoteRecords.size());
+      }
     }
 
-    // castVoteRecords will contain all cast vote records parsed by the reader
-    List<CastVoteRecord> castVoteRecords = new ArrayList<>();
-    // At each iteration of the following loop, we add records from another source file.
-    // source: index over config sources
-    for (RawElectionConfig.CVRSource source : config.rawConfig.cvrFileSources) {
-      Logger.info("Reading CVR file: %s (provider: %s)", source.filePath, source.provider);
-      // reader: read input file into a list of cast vote records
-      CVRReader reader = new CVRReader();
-      reader.parseCVRFile(
-          source.filePath,
-          source.firstVoteColumnIndex,
-          source.idColumnIndex,
-          source.precinctColumnIndex,
-          config
-      );
-      // add records to the master list
-      castVoteRecords.addAll(reader.castVoteRecords);
+    if (numInvalidSources > 0) {
```

Comment:
Merge this into the above check for `numInvalidSources > 0`.

Reply:
oops